heartbeat/monitors/active/http/http_test.go

```diff
@@ -47,59 +48,198 @@ func checkServer(t *testing.T, handlerFunc http.HandlerFunc) (*httptest.Server,
+	http.StatusPartialContent,
+	http.StatusMultiStatus,
+	http.StatusAlreadyReported,
+	http.StatusIMUsed,
+	// 3xx
+	http.StatusMultipleChoices,
+	http.StatusMovedPermanently,
+	http.StatusFound,
+	http.StatusSeeOther,
+	http.StatusNotModified,
+	http.StatusUseProxy,
+	http.StatusTemporaryRedirect,
+	http.StatusPermanentRedirect,
+}
+
+var downStatuses = []int{
+	//4xx
+	http.StatusBadRequest,
+	http.StatusUnauthorized,
+	http.StatusPaymentRequired,
+	http.StatusForbidden,
+	http.StatusNotFound,
+	http.StatusMethodNotAllowed,
+	http.StatusNotAcceptable,
+	http.StatusProxyAuthRequired,
+	http.StatusRequestTimeout,
+	http.StatusConflict,
+	http.StatusGone,
+	http.StatusLengthRequired,
+	http.StatusPreconditionFailed,
+	http.StatusRequestEntityTooLarge,
+	http.StatusRequestURITooLong,
+	http.StatusUnsupportedMediaType,
+	http.StatusRequestedRangeNotSatisfiable,
+	http.StatusExpectationFailed,
+	http.StatusTeapot,
+	http.StatusUnprocessableEntity,
+	http.StatusLocked,
+	http.StatusFailedDependency,
+	http.StatusUpgradeRequired,
+	http.StatusPreconditionRequired,
+	http.StatusTooManyRequests,
+	http.StatusRequestHeaderFieldsTooLarge,
+	http.StatusUnavailableForLegalReasons,
+	//5xx,
+	http.StatusInternalServerError,
+	http.StatusNotImplemented,
+	http.StatusBadGateway,
+	http.StatusServiceUnavailable,
+	http.StatusGatewayTimeout,
+	http.StatusHTTPVersionNotSupported,
+	http.StatusVariantAlsoNegotiates,
+	http.StatusInsufficientStorage,
+	http.StatusLoopDetected,
+	http.StatusNotExtended,
+	http.StatusNetworkAuthenticationRequired,
+}
+
+func TestUpStatuses(t *testing.T) {
+	for _, status := range upStatuses {
+		status := status
+		t.Run(fmt.Sprintf("Test OK HTTP status %d", status), func(t *testing.T) {
+			server, event := checkServer(t, hbtest.HelloWorldHandler(status))
+			port, err := hbtest.ServerPort(server)
+			require.NoError(t, err)
+
+			mapvaltest.Test(
+				t,
+				mapval.Strict(mapval.Compose(
+					hbtest.MonitorChecks("http@"+server.URL, server.URL, "127.0.0.1", "http", "up"),
+					hbtest.RespondingTCPChecks(port),
+					respondingHTTPChecks(server.URL, status),
+				)),
+				event.Fields,
+			)
+		})
+	}
+}
+
+func TestDownStatuses(t *testing.T) {
+	for _, status := range downStatuses {
+		status := status
+		t.Run(fmt.Sprintf("test down status %d", status), func(t *testing.T) {
+			server, event := checkServer(t, hbtest.HelloWorldHandler(status))
+			port, err := hbtest.ServerPort(server)
+			require.NoError(t, err)
+
+			errMsgCheck := mapval.IsStringContaining(fmt.Sprintf("%d", status))
+
+			mapvaltest.Test(
+				t,
+				mapval.Strict(mapval.Compose(
+					hbtest.MonitorChecks("http@"+server.URL, server.URL, "127.0.0.1", "http", "down"),
+					hbtest.RespondingTCPChecks(port),
+					respondingHTTPChecks(server.URL, status),
+					hbtest.ErrorChecks(errMsgCheck, "validate"),
+				)),
+				event.Fields,
+			)
+		})
+	}
+}
+
+func TestConnRefusedJob(t *testing.T) {
+	ip := "127.0.0.1"
+	port := uint16(freeport.GetPort())
```

Comment:
I wonder if you need a library here or you could just set :0 as port and it would automatically give you an available port.

Reply:
So, yes and no. Yes, that's exactly what freeport does, but if we didn't use it, we'd have to write it ourselves. The logic is pretty simple if you look at [freeport.go](https://github.com/phayes/freeport/blob/master/freeport.go), but I guess I don't see the point in rewriting it ourselves.

As a reminder, I'm essentially doing that in most tests, freeport is useful because I want an *unused* port, that will give me a TCP RST packet so I can see what the response is to a conn refused.

In the regular, server is up case, I leave the port unspecified and retrieve from the server object the port it decided to use.

Reply:
freeport only works because the net package uses `SO_REUSEADDR`. It has potential races, but for testing the approach is good enough I'd say.

Still, I'm no big fan if single-short-function dependencies ;)
It would be nice to have a more low-level common test support lib, providing this kind of function itself.

Reply:
+1 on not adding a dependency for it. Had a look at Freeport and it is only 15 lines of code: https://github.com/phayes/freeport/blob/master/freeport.go#L8

To move forward: we could keep it in this PR and remove it in a follow up PR where we introduce our own helper?